module/input/SpeechIntent/src/SpeechIntent.cpp

```diff
@@ -0,0 +1,391 @@
+            root_element = root_element->next;
+        }
+        free(root);
+
+        return intent;
+    }
+
+    std::unique_ptr<SpeechIntentMessage> parse_voice2json_json(std::string str) {
+        return parse_voice2json_json(str.data(), str.size());
+    }
+
+    // TODO: there no way of testing this for the moment.
+    // Use this reference example with "snd_pcm_poll_descriptors"
+    // Which returns a file handle we can wait on.
+    // https://gist.github.com/albanpeignier/104902
+    /*
+    bool start_microphone_capture() {
+        const char *device_name = "default";
+
+        int err = 0;
+        snd_pcm_t *capture_handle;
+        err = snd_pcm_open(&capture_handle, device_name, SND_PCM_STREAM_CAPTURE, 0);
+        if(err < 0) {
+            return false
+        }
+
+
+
+        return true;
+    }
+    */
+
+    // handles: out parameter containing the stdin, stdout, stderr handles from the
+    // spawned process
+    // first char * in args array refers to the process name
+    // args need to be null terminated. meaning that the last member of the should be null
+    bool spawn_process(SpawnedProcess& process, char** args) {
+        bool success = false;
+
+        pid_t pid                          = 0;
+        int stdout_pipe[2]                 = {};
+        int stderr_pipe[2]                 = {};
+        int stdin_pipe[2]                  = {};
+        posix_spawn_file_actions_t actions = {};
+        bool actions_inited                = false;
+
+        char* path = getenv("PATH");
+
+        size_t len     = strlen(path) + 5 + 1;
+        char* env_path = (char*) malloc(len);
+        int written    = sprintf(env_path, "PATH=%s", path);
+        assert(written > 0);
+        env_path[written] = 0;
+        
+        char* const envp[] = {(char*) "KALDI_DIR=",  // KALDI_DIR env var must be blank!
+                              (char*) env_path,
+                              NULL};
+
+        if (!((pipe(stdout_pipe) == -1) || (pipe(stderr_pipe) == -1) || (pipe(stdin_pipe) == -1))) {
+            posix_spawn_file_actions_init(&actions);
+            actions_inited = true;
+
+            posix_spawn_file_actions_addclose(&actions, stdout_pipe[0]);
+            posix_spawn_file_actions_addclose(&actions, stderr_pipe[0]);
+            posix_spawn_file_actions_addclose(&actions, stdin_pipe[1]);
+
+            posix_spawn_file_actions_adddup2(&actions, stdout_pipe[1], STDOUT_FILENO);
+            posix_spawn_file_actions_adddup2(&actions, stderr_pipe[1], STDERR_FILENO);
+            posix_spawn_file_actions_adddup2(&actions, stdin_pipe[0], STDIN_FILENO);
+
+            posix_spawn_file_actions_addclose(&actions, stdout_pipe[1]);
+            posix_spawn_file_actions_addclose(&actions, stderr_pipe[1]);
+            posix_spawn_file_actions_addclose(&actions, stdin_pipe[0]);
+
+            if (posix_spawnp(&pid, (char const*) args[0], &actions, 0, (char* const*) args, envp) == 0) {
+                success = true;
```

Comment:
I know this technically works, but feels clunky for some reason, there's probably a better way to cleanup and return?

Reply:
What looks clunky to you? All 3 return "paths" - spawn process success, pipe creation failure and spawn process failure need to cleanup those resources ? How should we proceed?